File: Computer Science/Hardik Ahuja/Task 2/mergesort.py
def merge(first, second, final):
	lf = len(first)
	ls = len(second)
	i=0
	j=0
	k=0
	while(i<lf and j<ls):
		if first[i]<=second[j]:
			final[k] = first[i]
			k+=1
			i+=1
		else:
			final[k] = second[j]
			k+=1
			j+=1
	while(i<lf):
		final[k] = first[i]
		k+=1
		i+=1
	while(j<ls):
		final[k] = second[j]
		k+=1
		j+=1
	#print(final)

def mergesort(list):
	n = len(list)
	left = [0]*(n//2)
	right = [0]*(n-(len(left)))
	#print(left)
	#print(right)
	if n<2:
		#print(list)
		return 0
	print("Initial ",list)
	for i in range(n//2):
		#print("i",i)
		left[i] = list[i]
	for j in range(n//2,n):
		#print("j",j)
		right[j-(n//2)] = list[j]
	#print("lleefftt")
	mergesort(left)
	#print(left)
	#print("rriigghhtt")
	mergesort(right)
	merge(left, right, list)
	print("Sorted ",list)

#merge([1,2,4,6],[3,5,7,8],[1,4,3,2,5,6,7,8])
list=[8,7,2,5,6,1,4,3]
mergesort(list)
print("Final Sorted list ",list)

File: Computer Science/Atishi Miral/Sorting Algorithms/Merge Sort.cpp
#include<iostream>
using namespace std;
 

//Using divide and conquer strategy

void merge(int *a, int l, int m, int r)
{
    int n1= m-l+1; 	//Length of Left subaaray
    int n2= r-m;	//Length of Right subarray
 
    int L[n1], R[n2];	// Left side and Right side of the sorted array
 
    for (int i=0; i<n1; i++)
        L[i] = a[l + i];
    for (int j=0; j<n2; j++)
        R[j] = a[m + 1+ j];
 
    int x=0; // Initial index of first subarray
    int y=0; // Initial index of second subarray
    int z=l; // Initial index of sorted subarray
    while (x < n1 && y < n2)
    {
        if (L[x] <= R[y])
        {
            a[z] = L[x];
            x++;
        }
        else
        {
            a[z] = R[y];
            y++;
        }
        z++;
    }
 
    while (x < n1)
    {
        a[z] = L[x];
        x++;
        z++;
    }
 
    while (y < n2)
    {
        a[z] = R[y];
        y++;
        z++;
    }
}
 

void mergeSort(int *a, int l, int r)
{
    if (l<r) //l is for left index and r is for right index of the array
    {

        int m= (l+r)/2; // m=mid value; l+(r-l)/2 can be used for large values of h and l
 
        mergeSort(a, l, m);
        mergeSort(a, m+1, r);
 
        merge(a, l, m, r);
    }
}
 
void print(int *a, int len)
{
    for (int i=0; i<len; i++)
        cout<<a[i]<<"\t";
    cout<<endl;
}
 
int main()
{
    int n, arr[20];
    cout<<"Enter size of array: ";	cin>>n; cout<<endl<<"Enter elements:"<<endl;
    for (int i=0; i<n; i++)
    	cin>>arr[i];
 
    mergeSort(arr, 0, n-1);
 
    cout<<"Sorted array:"<<endl;
    print(arr, n);
    return 0;
}
